keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -389,38 +382,100 @@ type IngressSpec struct {
+priority. The controller MAY emit a warning event if the user sets conflicting
+(different) values for the annotation and field.
 
-...
+#### IngressClass Resource
 
-// IngressClass represents the class of the Ingress, referenced by the
-// ingress.spec. IngressClass will be a non-namespaced Cluster resource.
+Additionally, we propose adding an `IngressClass` resource to provide additional
+data about the Ingress class. This resource will be non-namespaced. The
+IngressClass resource is an optional way to provide additional configuration for
+a specific class of Ingress. This allows us to evolve the API to express
+concepts such as levels of service associated with a given Ingress controller.
+
+The name of this IngressClass resource will be tied to any Ingresses with the
+same value for the class field or annotation. An IngressClass resource can exist
+without any Ingresses referencing it, and an Ingress can have a class value that
+does not correspond with an IngressClass resource.
+
+```golang
+// IngressClass represents the class of the Ingress, referenced by the Ingress
+// Spec.
 type IngressClass struct {
   metav1.TypeMeta
   metav1.ObjectMeta
 
+  // Spec is the desired state of the IngressClass.
+  // More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+  // +optional
+  Spec IngressClassSpec
+
+  // Status is the current state of the IngressClass.
+  // More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#spec-and-status
+  // +optional
+  Status IngressClassStatus
+}
+
+// IngressClassSpec provides information about the class of an Ingress.
+type IngressClassSpec struct {
   // Controller is responsible for handling this class. This should be
   // specified as a domain-prefixed path, e.g. "acme.io/ingress-controller".
-  //
   // This allows for different "flavors" that are controlled by the same
-  // controller. For example, you may have different Parameters for
-  // the same implementing controller.
+  // controller. For example, you may have different Parameters for the same
+  // implementing controller.
   Controller string
 
-  // Parameters is a link to a custom resource configuration for
-  // the controller. This is optional if the controller does not
-  // require extra parameters.
-  //
+  // Parameters is a link to a custom resource configuration for the
+  // controller. This is optional if the controller does not require extra
+  // parameters.
   // +optional
-  Parameters *TypeLocalObjectReference
+  Parameters *api.TypedLocalObjectReference
 }
-```
 
-##### Interoperability with previous annotation
+// IngressClassStatus provides information about the status of an IngressClass.
+type IngressClassStatus struct {
+  // Conditions is the current status from the controller for provisioning a
+  // class of Gateways with the given set of Parameters.
+  // +optional
+  Conditions []IngressClassCondition
```

Comment:
Why is this status on the Class rather than on the Ingresses themselves?  None of the other *Class types follow this.

Reply:
This is mirroring what we're working towards with GatewayClass. The idea is that implementations of the Ingress or Gateway API would like to add status on Class resources to indicate whether or not configuration at that level is valid. This is not meant to say anything about the Ingress resources, more the specific configuration of this class and if it's valid.

Reply:
Thinking about this more -- it is likely that the controller may not be able to parse and give the full set of errors w/out an associated ingress / gateway definition. For synchronous validation, the controllers can introduce webhooks to cover that case. Maybe we should leave this out for now? (status)

Reply:
@bowei That makes sense, I've removed Status for now. We can revisit this after we determine what makes sense on GatewayClass.

Reply:
Yeah, I don't think the class is the correct place for this status.